memory access violation in master

```bash
make test
npm test

> @mapbox/node-cpp-skel@0.1.0 test /home/artem/projects/vtcomposite
> tape test/*.test.js

TAP version 13
# [composite] success: buffer size stays the same when no compositing needed
[buffer size] cpp pre vtzero: 101067
[buffer size] cpp output:101067
ok 1 should be falsy
ok 2 same size
# [composite] success compositing - same layer name, same features, same zoom
free(): invalid pointer
Aborted (core dumped)
npm ERR! Test failed.  See above for more details.
Makefile:90: recipe for target 'test' failed
make: *** [test] Error 1
```

The above triggered by:  https://github.com/mapbox/vtcomposite/blob/master/test/vtcomposite.test.js#L43

--- a/bench/vtcomposite.bench.js
+++ b/bench/vtcomposite.bench.js
@@ -16,10 +16,10 @@ process.env.UV_THREADPOOL_SIZE = argv.concurrency;
 
 var fs = require('fs');
 var path = require('path');
-var bytes = require('bytes');  
+var bytes = require('bytes');
 var assert = require('assert')
 var d3_queue = require('d3-queue');
-var module = require('../lib/index.js');
+var composite = require('../lib/index.js');
 var queue = d3_queue.queue();
 var bufferSF = fs.readFileSync(path.resolve(__dirname+'/../node_modules/@mapbox/mvt-fixtures/real-world/sanfrancisco/15-5238-12666.mvt'));
 var bufferNepal = fs.readFileSync(path.resolve(__dirname+'/../node_modules/@mapbox/mvt-fixtures/real-world/nepal/13-6036-3426.mvt'));
@@ -31,7 +31,7 @@ var tiles = [
 
 var zxy = {z:15, x:5238, y:12666};
 
-var track_mem = argv.mem ? true : false; 
+var track_mem = argv.mem ? true : false;
 var runs = 0;
 var memstats = {
   max_rss:0,
@@ -41,7 +41,7 @@ var memstats = {
 
 function run(cb) {
   // running our module
-  module.composite(tiles, zxy, {}, function(err, result) {
+  composite(tiles, zxy, {}, function(err, result) {
       if (err) {
         return cb(err);
       }
@@ -57,7 +57,7 @@ function run(cb) {
 }
 
 // Start monitoring time before async work begins within the defer iterator below.
-// AsyncWorkers will kick off actual work before the defer iterator is finished, 
+// AsyncWorkers will kick off actual work before the defer iterator is finished,
 // and we want to make sure we capture the time of the work of that initial cycle.
 var time = +(new Date());
 
@@ -79,7 +79,7 @@ queue.awaitAll(function(error) {
   // number of milliseconds per iteration
     var rate = runs/(time/1000);
     console.log('Benchmark speed: ' + rate.toFixed(0) + ' runs/s (runs:' + runs + ' ms:' + time + ' )');
-  
+
     if (track_mem) {
       console.log('Benchmark peak mem (max_rss, max_heap, max_heap_total): ', bytes(memstats.max_rss), bytes(memstats.max_heap), bytes(memstats.max_heap_total));
     } else {

--- a/package.json
+++ b/package.json
@@ -26,7 +26,7 @@
     "node-pre-gyp"
   ],
   "devDependencies": {
-    "@mapbox/mvt-fixtures": "^3.3.1",
+    "@mapbox/mvt-fixtures": "^3.4.0",
     "@mapbox/vector-tile": "^1.3.1",
     "@mapbox/vtinfo": "^0.1.2",
     "aws-sdk": "^2.4.7",

--- a/src/extract_geometry.hpp
+++ b/src/extract_geometry.hpp
@@ -3,7 +3,8 @@
 #include <mapbox/geometry.hpp>
 #include <vtzero/vector_tile.hpp>
 
-namespace vtile { namespace detail {
+namespace vtile {
+namespace detail {
 
 template <typename CoordinateType>
 struct point_geometry_handler

--- a/src/vtcomposite.cpp
+++ b/src/vtcomposite.cpp
@@ -157,7 +157,7 @@ struct CompositeWorker : Nan::AsyncWorker
                     std::cerr << "Invalid tile composite request" << std::endl;
                 }
             }
-            std::string & tile_buffer = *output_buffer_.get();
+            std::string& tile_buffer = *output_buffer_.get();
             builder.serialize(tile_buffer);
         }
         catch (std::exception const& e)
@@ -168,18 +168,18 @@ struct CompositeWorker : Nan::AsyncWorker
 
     void HandleOKCallback() override
     {
-        std::string & tile_buffer = *output_buffer_.get();
+        std::string& tile_buffer = *output_buffer_.get();
         Nan::HandleScope scope;
         const auto argc = 2u;
         v8::Local<v8::Value> argv[argc] = {
             Nan::Null(),
             Nan::NewBuffer(&tile_buffer[0],
                            static_cast<std::uint32_t>(tile_buffer.size()),
-                           [](char *, void * hint) {
-                              delete reinterpret_cast<std::string*>(hint);
+                           [](char*, void* hint) {
+                               delete reinterpret_cast<std::string*>(hint);
                            },
-                           output_buffer_.release()
-                          ).ToLocalChecked()};
+                           output_buffer_.release())
+                .ToLocalChecked()};
 
         // Static cast done here to avoid 'cppcoreguidelines-pro-bounds-array-to-pointer-decay' warning with clang-tidy
         callback->Call(argc, static_cast<v8::Local<v8::Value>*>(argv), async_resource);

--- a/src/vtcomposite.cpp
+++ b/src/vtcomposite.cpp
@@ -84,7 +84,7 @@ struct CompositeWorker : Nan::AsyncWorker
     CompositeWorker(std::unique_ptr<BatonType> baton_data, Nan::Callback* cb)
         : Base{cb},
           baton_data_{std::move(baton_data)},
-          output_buffer_{} {}
+          output_buffer_{std::make_unique<std::string>()} {}
 
     void Execute() override
     {
@@ -114,7 +114,6 @@ struct CompositeWorker : Nan::AsyncWorker
                         tile_view = tile_obj->data;
                     }
 
-                    std::cout << "[buffer size] cpp pre vtzero: " << tile_view.size() << std::endl;
                     int zoom_factor = 1 << (target_z - tile_obj->z);
                     vtzero::vector_tile tile{tile_view};
                     while (auto layer = tile.next_layer())
@@ -158,8 +157,8 @@ struct CompositeWorker : Nan::AsyncWorker
                     std::cerr << "Invalid tile composite request" << std::endl;
                 }
             }
-            builder.serialize(output_buffer_);
-            std::cerr << "[buffer size] cpp output:" << output_buffer_.size() << std::endl;
+            std::string & tile_buffer = *output_buffer_.get();
+            builder.serialize(tile_buffer);
         }
         catch (std::exception const& e)
         {
@@ -169,18 +168,25 @@ struct CompositeWorker : Nan::AsyncWorker
 
     void HandleOKCallback() override
     {
+        std::string & tile_buffer = *output_buffer_.get();
         Nan::HandleScope scope;
         const auto argc = 2u;
         v8::Local<v8::Value> argv[argc] = {
             Nan::Null(),
-            Nan::NewBuffer(&output_buffer_[0], static_cast<std::uint32_t>(output_buffer_.size())).ToLocalChecked()};
+            Nan::NewBuffer(&tile_buffer[0],
+                           static_cast<std::uint32_t>(tile_buffer.size()),
+                           [](char *, void * hint) {
+                              delete reinterpret_cast<std::string*>(hint);
+                           },
+                           output_buffer_.release()
+                          ).ToLocalChecked()};
 
         // Static cast done here to avoid 'cppcoreguidelines-pro-bounds-array-to-pointer-decay' warning with clang-tidy
         callback->Call(argc, static_cast<v8::Local<v8::Value>*>(argv), async_resource);
     }
 
     std::unique_ptr<BatonType> const baton_data_;
-    std::string output_buffer_;
+    std::unique_ptr<std::string> output_buffer_;
 };
 
 NAN_METHOD(composite)

--- a/src/zoom_coordinates.hpp
+++ b/src/zoom_coordinates.hpp
@@ -7,7 +7,8 @@
 //#include <mapbox/geometry/algorithms/detail/boost_adapters.hpp>
 //#include <boost/geometry/algorithms/for_each.hpp>
 
-namespace vtile { namespace detail {
+namespace vtile {
+namespace detail {
 
 template <typename Point>
 struct zoom_coordinates
@@ -16,13 +17,13 @@ struct zoom_coordinates
     explicit zoom_coordinates(coordinate_type factor)
         : factor_(factor) {}
 
-    inline void operator()(Point & p) const
+    inline void operator()(Point& p) const
     {
         p.x *= factor_;
         p.y *= factor_;
     }
     coordinate_type factor_;
 };
-}
+} // namespace detail
 
-}
+} // namespace vtile

--- a/test/vtcomposite.test.js
+++ b/test/vtcomposite.test.js
@@ -9,19 +9,8 @@ var pbf = require('pbf');
 
 function vtinfo(buffer) {
   var tile = new vt(new pbf(buffer));
-  var info = {
-    layers : []
-  };
-  Object.keys(tile.layers).forEach(function(k) {
-    var lay = tile.layers[k];
-    info.layers.push({
-      name:k,
-      features:lay.length
-    })
-  });
-  return info;
+  return tile;
 }
-
 
 var bufferSF = fs.readFileSync(path.resolve(__dirname+'/../node_modules/@mapbox/mvt-fixtures/real-world/sanfrancisco/15-5238-12666.mvt'));
 
@@ -40,10 +29,10 @@ test('[composite] success: buffer size stays the same when no compositing needed
 });
 
 test('[composite] success compositing - same layer name, same features, same zoom', function(assert) {
-  const singlePointBuffer = mvtFixtures.get('017').buffer; 
+  const singlePointBuffer = mvtFixtures.get('017').buffer;
 
   const tiles = [
-    {buffer: singlePointBuffer, z:15, x:5238, y:12666}, 
+    {buffer: singlePointBuffer, z:15, x:5238, y:12666},
     {buffer: singlePointBuffer, z:15, x:5238, y:12666}
   ];
 
@@ -54,21 +43,21 @@ test('[composite] success compositing - same layer name, same features, same zoo
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const outputInfo = vtinfo(vtBuffer);
 
-    assert.equal(outputInfo.layers[0].name, 'hello');
-    assert.equal(outputInfo.layers.length, 1);
-    assert.equal(outputInfo.layers[0].features, 1);
+    assert.ok(outputInfo.layers.hello, 'hello', 'expected layer name');
+    assert.equal(Object.keys(outputInfo.layers).length, 1, 'expected number of layers');
+    assert.equal(outputInfo.layers.hello.length, 1, 'expected number of features');
     assert.notOk(err);
     assert.end();
   });
 });
 
 test('[composite] success compositing - same layer name, different features, same zoom', function(assert) {
-  const singlePointBuffer = mvtFixtures.get('017').buffer; 
-  const singleLineStringBuffer = mvtFixtures.get('018').buffer; 
+  const buffer1 = mvtFixtures.get('059').buffer; // mud lake
+  const buffer2 = mvtFixtures.get('060').buffer; // crater lake
 
   const tiles = [
-    {buffer: singleLineStringBuffer, z:15, x:5238, y:12666},
-    {buffer: singlePointBuffer, z:15, x:5238, y:12666}
+    {buffer: buffer1, z:15, x:5238, y:12666}, // this layer wins
+    {buffer: buffer2, z:15, x:5238, y:12666}
   ];
 
   const zxy = {z:15, x:5238, y:12666};
@@ -76,18 +65,18 @@ test('[composite] success compositing - same layer name, different features, sam
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const outputInfo = vtinfo(vtBuffer);
 
-    assert.equal(outputInfo.layers[0].name, 'hello', 'returns layer hello');
-    assert.equal(outputInfo.layers.length, 1, 'return 1 layers');
-    assert.equal(outputInfo.layers[0].features, 1, 'returns 1 feature');
-    assert.equal(outputInfo.layers[0].line_features, 1, 'keeps feature from first layer');
+    assert.ok(outputInfo.layers.water, 'returns layer water');
+    assert.equal(Object.keys(outputInfo.layers).length, 1, 'return 1 layers');
+    assert.equal(outputInfo.layers.water.length, 1, 'only has one feature');
+    assert.equal(outputInfo.layers.water.feature(0).properties.name, 'mud lake', 'expected feature');
     assert.notOk(err);
     assert.end();
   });
 });
 
 test('[composite] success compositing - different layer name, different features, same zoom', function(assert) {
-  const buffer1 = mvtFixtures.get('017').buffer; 
-  const buffer2 = mvtFixtures.get('053').buffer; 
+  const buffer1 = mvtFixtures.get('017').buffer;
+  const buffer2 = mvtFixtures.get('053').buffer;
 
   const tiles = [
     {buffer: buffer1, z:15, x:5238, y:12666},
@@ -99,9 +88,9 @@ test('[composite] success compositing - different layer name, different features
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const outputInfo = vtinfo(vtBuffer);
 
-    assert.equal(outputInfo.layers[0].name, 'hello', 'returns layer hello');
-    assert.equal(outputInfo.layers[1].name, 'clipped-square', 'returns layer hello');
-    assert.equal(outputInfo.layers.length, 2, 'return 2 layers');
+    assert.ok(outputInfo.layers.hello, 'expected layer name');
+    assert.ok(outputInfo.layers['clipped-square'], 'expected layer name');
+    assert.equal(Object.keys(outputInfo.layers).length, 2, 'expected number of layers');
     assert.notOk(err);
     assert.end();
   });
@@ -123,12 +112,12 @@ test('[composite] success: compositing single gzipped VT', function(assert) {
   });
 });
 
-// TEST - check vt is within target 
+// TEST - check vt is within target
 
 test('[composite] failure? discards layer that is not within target', function(assert) {
   // TBD whether it's a failure or silently discards the tile that isn't in the target
-  const buffer1 = mvtFixtures.get('017').buffer; 
-  const buffer2 = mvtFixtures.get('053').buffer; 
+  const buffer1 = mvtFixtures.get('017').buffer;
+  const buffer2 = mvtFixtures.get('053').buffer;
 
   const tiles = [
     {buffer: buffer1, z:15, x:5238, y:12666},
@@ -139,8 +128,8 @@ test('[composite] failure? discards layer that is not within target', function(a
 
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const info = vtinfo(vtBuffer)
-    assert.equal(info.layers.length, 1, 'discards layer that is not within target');
-    assert.equal(info.layers[0].name, 'hello', 'layer within target created into output buffer');
+    assert.equal(Object.keys(info.layers).length, 1, 'discards layer that is not within target');
+    assert.ok(info.layers.hello, 'layer within target created into output buffer');
     assert.notOk(err);
     assert.end();
   });
@@ -148,20 +137,20 @@ test('[composite] failure? discards layer that is not within target', function(a
 
 test('[composite] failure? tile does not intersect target zoom level ', function(assert) {
   // TBD whether it's a failure or silently discards the tile that isn't in the target
-  const buffer1 = mvtFixtures.get('017').buffer; 
-  const buffer2 = mvtFixtures.get('053').buffer; 
+  const buffer1 = mvtFixtures.get('017').buffer;
+  const buffer2 = mvtFixtures.get('053').buffer;
 
   const tiles = [
     {buffer: buffer1, z:7, x:19, y:44},
-    {buffer: buffer2, z:6, x:10, y:22} //this tile does not intersect target zoom level 
+    {buffer: buffer2, z:6, x:10, y:22} //this tile does not intersect target zoom level
   ];
 
   const zxy = {z:7, x:19, y:44};
 
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const info = vtinfo(vtBuffer)
-    assert.equal(info.layers.length, 1, 'discards layer that is not within target');
-    assert.equal(info.layers[0].name, 'hello', 'layer within target created into output buffer');
+    assert.equal(Object.keys(info.layers).length, 1, 'discards layer that is not within target');
+    assert.ok(info.layers.hello, 'layer within target created into output buffer');
     assert.notOk(err);
     assert.end();
   });
@@ -169,20 +158,20 @@ test('[composite] failure? tile does not intersect target zoom level ', function
 
 test('[composite] failure? tile with zoom level higher than requested zoom is discarded', function(assert) {
   // TBD whether it's a failure or silently discards the tile that isn't in the target
-  const buffer1 = mvtFixtures.get('017').buffer; 
-  const buffer2 = mvtFixtures.get('053').buffer; 
+  const buffer1 = mvtFixtures.get('017').buffer;
+  const buffer2 = mvtFixtures.get('053').buffer;
 
   const tiles = [
-    {buffer: buffer2, z:7, x:10, y:22}, 
-    {buffer: buffer2, z:6, x:10, y:22} 
+    {buffer: buffer2, z:7, x:10, y:22},
+    {buffer: buffer2, z:6, x:10, y:22}
   ];
 
   const zxy = {z:6, x:10, y:22};
 
   composite(tiles, zxy, {}, (err, vtBuffer) => {
     const info = vtinfo(vtBuffer)
-    assert.equal(info.layers.length, 1, 'discards layer that is not within target');
-    assert.equal(info.layers[0].name, 'clipped-square', 'layer within target created into output buffer');
+    assert.equal(Object.keys(info.layers).length, 1, 'discards layer that is not within target');
+    assert.ok(info.layers['clipped-square'], 'layer within target created into output buffer');
     assert.notOk(err);
     assert.end();
   });
@@ -192,7 +181,7 @@ test('[composite] failure? tile with zoom level higher than requested zoom is di
 
 test('failure: fails without callback function', assert => {
   try {
-    
+
 composite();
   } catch(err) {
     assert.ok(/last argument must be a callback function/.test(err.message), 'expected error message');

